dlib/te.h

```diff
@@ -0,0 +1,449 @@
+// Copyright (C) 2022  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_TYPE_ERASURE_H_
+#define DLIB_TYPE_ERASURE_H_
+
+#include <type_traits>
+#include <utility>
+#include <typeindex>
+#include <new>
+#include <memory>
+
+namespace dlib
+{
+    namespace te
+    {
+        template<class T, bool null_after_move>
+        struct managed_move_pointer
+        {
+            managed_move_pointer(T* ptr_) : ptr{ptr_} {}
+            managed_move_pointer()                                          = default;
+            managed_move_pointer(const managed_move_pointer&)               = default;
+            managed_move_pointer& operator=(const managed_move_pointer&)    = default;
+
+            managed_move_pointer(managed_move_pointer&& other) noexcept 
+            : ptr{other.ptr}
+            {
+                if (null_after_move)
+                    other.ptr = nullptr;
+            }
+
+            managed_move_pointer& operator=(managed_move_pointer&& other) noexcept
+            {
+                if (this != &other)
+                {
+                    ptr = other.ptr;
+                    if (null_after_move)
+                        other.ptr = nullptr;
+                }
+                return *this;
+            }
+
+            operator bool() const       { return ptr != nullptr; }
+            const T* operator->() const { return ptr; }
+            T*       operator->()       { return ptr; }     
+
+            T* ptr = nullptr;
```

Comment:
Totally don't have public member variables.  Especially with the other stuff.  Got to protect the class invariants.

Reply:
yep